### DISCUSSION TITLE
ERROR: Too many messages queued

### DISCUSSION THREAD

**Discussion by dgbalharrie** [bertmelis/espMqttClient#90] [General]
On an ESP32 Dev board, I am getting "ERROR: Too many messages queued" when I publish a message.

Can't find a reference to this in the code or docs, so not sure what I need to do.

**Comment by bertmelis** [comment]
This doesn't seem to be a message from this library. Do you have any code or log you can share?

**Reply by dgbalharrie** [reply]
Thank you for the reply. I think you are right now... the message was appearing after I made some changes to topic names and it appeared after publish message in my trace...
I am using AsyncWebServer with socketio messages to a webpage. Closing the webpage and the error is no longer there.

**Comment by dgbalharrie** [comment]
The Mqtt message did seem to provoke the error, however, I was sending a two socketio messages once a second to the webpage. I revised this to once a minute and I now don't get the error. I realise this error was coming from the AsyncTCP library and I note others have observed the error. Looking at the AsyncTCP docs, there isn't a way to increase the queue other than changing some #defines. I know the error isn't in this library, but just noting my experience.

**Reply by bertmelis** [reply]
2 messages per second doesn't seem excessive to me. Maybe you're sending characters/bytes one by one?

**Comment by dgbalharrie** [comment]
I agree it doesn’t seem excessive.
No I am sending a char string as a message. The socketio messages are a json pair… the mqtt, a simple digit or digit pair, though I was wondering if the topics are too long. For example "mainControl/heating/temperature/CurrentTargetTemp"

However, it does seem to be a reported issue. https://github.com/me-no-dev/ESPAsyncWebServer/issues/1158


I am not using ws.availableForWriteAll() as it didn’t appear in the docs/examples I saw… maybe I should try it.

I was using a MQTT library that is in many of the tutorials online. The library worked fine when I was using two ESP32 devices talking to each other, but when I brought up a third, the devices seemed to go crazy with a network IO error, causing them to reboot. So far your library has been working fine, despite the two many messages queued error, which isn’t from you library. So far I have been happy with your library

**Reply by bertmelis** [reply]
This library works a bit different. It queues the messages and sends when the underlying network is available for writing. By putting the messages in a queue, QoS > 0 is possible which isn't available in PubSubClient for example. Disadvantage is memory usage and possible memory defragmentation. but the ESP8266 and ESP32 have plenty so it shouldn't be a problem.

To really understand you problem, you will need to share more of your design.

If I understand correctly: you have a webserver on the esp32 and serve a webpage that connects through websockets back to the esp? Messages from MQTT are forwarded to websockets and vice-versa?

How any websocket clients (eg browser windows) are you serving?

**Comment by dgbalharrie** [comment]
I am serving one web client. I have QoS set to 0 at the moment.

Do you have an account on gitlab? I can add you to the project to look at the code.
